python/dgl/batched_graph.py

```diff
@@ -751,20 +760,199 @@ def _max_on(graph, typestr, feat):
 
     if isinstance(graph, BatchedDGLGraph):
         batch_num_objs = getattr(graph, batch_num_objs_attr)
-        max_readout_list = []
-        first = 0
-        for num_obj in batch_num_objs:
+        max_n_objs = max(batch_num_objs)
+        index, mask = [], []
+        for i, num_obj in enumerate(batch_num_objs):
             if num_obj == 0:
-                max_readout_list.append(F.zeros(F.shape(feat)[1:],
-                                                F.dtype(feat),
-                                                F.context(feat)))
-                continue
-            max_readout_list.append(F.max(feat[first:first+num_obj], 0))
-            first += num_obj
-        return F.stack(max_readout_list, 0)
+                dgl_warning("Graph {} has zero {}, a zero tensor with the same shape"
+                            " would be returned at the corresponding row.".format(i, typestr))
+                mask.append(0.)
+            else:
+                mask.append(1.)
+            index.extend(range(i * max_n_objs, i * max_n_objs + num_obj))
+        dtype = F.dtype(feat)
+        ctx = F.context(feat)
+        mask = F.copy_to(F.tensor(mask), ctx)
+        index = F.copy_to(F.tensor(index), ctx)
+        val = F.reduce_min(feat) - 1
+        feat_ = F.zeros((len(batch_num_objs) * max_n_objs, *F.shape(feat)[1:]),
+                        dtype, ctx) + val
+        feat_ = F.scatter_row(feat_, index, feat)
+        feat_ = F.reshape(feat_, (len(batch_num_objs), max_n_objs, *F.shape(feat)[1:]))
+        return F.max(feat_, 1) * F.unsqueeze(mask, -1)
```

Comment:
Looks like we are padding the entries to the same length.  Could you add a comment (something like "pad the nodes/edges to the same length with -inf before taking a max")?

Also, why use `F.reduce_min(feat) - 1` instead of `-inf`?

An unimportant question: do we have a DGL solution that does things like `scatter_max`?

Reply:
The result of `-inf * 0` is `NaN`.

Reply:
Yes we can, create a pseudo graph with a summarization node connects to each node and call `update_all`. However, at current stage the tensor solution might be better.

Reply:
DONE.